src/lib.rs

```diff
@@ -23,389 +32,524 @@ unsafe impl Send for LDAP {}
+    fn ldap_search_ext_s(ldap: *mut LDAP, base: *const c_char, scope: c_int,
+                         filter: *const c_char, attrs: *const *const c_char,
+                         attrsonly: c_int, serverctrls: *mut *mut LDAPControl,
+                         clientctrls: *mut *mut LDAPControl, timeout: *mut timeval,
+                         sizelimit: c_int, res: *mut *mut LDAPMessage) -> c_int;
+    fn ldap_unbind_ext_s(ldap: *mut LDAP, sctrls: *mut *mut LDAPControl, cctrls: *mut *mut LDAPControl) -> c_int;
 }
 
+/// A typedef for an LDAPResponse type.
+///
+/// LDAP responses are organized as vectors of mached entities. Typically, each entity is
+/// represented as a map of attributes to list of values.
+///
+pub type LDAPResponse = Vec<HashMap<String,Vec<String>>>;
+
+
+/// A high level abstraction over the raw OpenLDAP functions.
+///
+/// A `RustLDAP` object hides raw OpenLDAP complexities and exposes a simple object that is
+/// created, configured, and queried. Methods that call underlying OpenLDAP calls that can fail
+/// will raise an `errors::LDAPError` with additional details.
+///
+/// Using a `RustLDAP` object is easy!
+///
 pub struct RustLDAP {
-	// Have the raw pointer to it so we can pass it into internal functions
-	ldap_ptr: *mut LDAP
+    /// A pointer to the underlying OpenLDAP object.
+    ldap_ptr: *mut LDAP,
 }
 
 impl Drop for RustLDAP {
+    fn drop(&mut self){
+        // Unbind the LDAP connection, making the C library free the LDAP*.
+        let rc = unsafe { ldap_unbind_ext_s(self.ldap_ptr, ptr::null_mut(), ptr::null_mut()) };
+
+        // Make sure it actually happened.
+        if rc != codes::results::LDAP_SUCCESS {
+            unsafe {
+                // Hopefully this never happens.
+                let raw_estr = ldap_err2string(rc as c_int);
+                panic!(CStr::from_ptr(raw_estr).to_owned().into_string().unwrap());
+            }
+        }
+    }
+}
 
-	fn drop(&mut self){
-
-		//unbind the LDAP connection, making the C library free the LDAP*
-		let rc = unsafe { ldap_unbind_ext_s(self.ldap_ptr, ptr::null_mut(), ptr::null_mut()) };
+/// A trait for types that can be passed as LDAP option values.
+///
+/// Underlying OpenLDAP implementation calls for option values to be passed in as *const c_void,
+/// while allowing values to be i32 or string. Using traits, we implement function overloading to
+/// handle i32 and string option value types.
+///
+pub trait LDAPOptionValue {
+    fn as_cvoid_ptr(&self) -> *const c_void;
+}
 
-		//make sure it actually happened
-		if rc != codes::results::LDAP_SUCCESS {
-			unsafe { //hopefully this never happens
-				let raw_estr = ldap_err2string(rc as c_int);
-				panic!(CStr::from_ptr(raw_estr).to_owned().into_string().unwrap());
-			}
+impl LDAPOptionValue for str {
+    fn as_cvoid_ptr(&self) -> *const c_void {
+        let string = CString::new(self).unwrap();
+        unsafe {
+            return mem::transmute(string.as_ptr());
```

Comment:
Okay, the rabbit hole goes deep, but I figured this out with some help.

The method `CString::into_raw` is what we want here, and transfers ownership to the C function, so no worry about use after free issue. You might be able to use that pattern in FFI-land elsewhere.

However, then the problem is when to free using `CString::from_raw`. Naturally, we'd opt to do this in `RustLDAP::set_option`, but this then forces `RustLDAP::set_option` to know i) that certain options are strings and ii) that the `LDAPOptionValue` trait sometimes allocates memory. Seems to be coupling the two objects together, which I don't like.

At this point, I might rewrite to change the signature of `LDAPOptionValue::as_cvoid_ptr(&self)` to be something like `LDAPOptioNValue::into_cvoid_ptr(&self, buf: *mut c_void)`, which forces the caller to provide a buffer we can use.